contents/handbook/engineering/session-replay/session-replay-architecture.md

```diff
@@ -0,0 +1,217 @@
+
+### Phase 1: Rust capture service
+**`rust/capture/src/v0_endpoint.rs`**
+- Receives POST to `/s/` endpoint
+- Validates session_id (rejects malformed/too long IDs)
+- Routes to replay-specific Kafka sink (`process_replay_events`)
+- Publishes to **`session_recording_snapshot_item_events`** topic (or `_overflow` topic if billing-limited)
+
+**Kafka sink** (`rust/capture/src/sinks/kafka.rs`):
+- Produces to primary topic: `KAFKA_SESSION_RECORDING_SNAPSHOT_ITEM_EVENTS`
+- Overflow topic: `KAFKA_SESSION_RECORDING_SNAPSHOT_ITEM_OVERFLOW`
+
+### Phase 2: Blob ingestion consumer (Node.js/TypeScript)
+**`plugin-server/src/main/ingestion-queues/session-recording-v2/`**
+
+**SessionRecordingIngester** consumes from Kafka and:
+1. **Parses** gzipped/JSON messages (`kafka/message-parser.ts`)
+2. **Batches** by session via **SessionBatchRecorder** 
+3. **Buffers** events in memory per session using **SnappySessionRecorder**:
+
+```ts
+    public recordMessage(message: ParsedMessageData): number {
+        if
+   ...
+   133| {
+            this.endDateTime = message.eventsRange.end
+        }
+
+        for (const [windowId, events] of Object.entries(message.eventsByWindowId)) {
+            for (const event of events) {
+                const serializedLine = JSON.stringify([windowId, event]) + '\n'
+                const chunk = Buffer.from(serializedLine)
+                this.uncompressedChunks.push(chunk)
+
+                const eventTimestamp = event.timestamp
+                const shouldComputeMetadata = eventPassesMetadataSwitchoverTest(
+                    eventTimestamp,
+                    this.metadataSwitchoverDate
+                )
+
+                if (shouldComputeMetadata) {
+                    // Store segmentation event for later use in active time calculation
+                    this.segmentationEvents.push(toSegmentationEvent(event))
+
+                    const eventUrl = hrefFrom(event)
+                    if (eventUrl) {
+                        this.addUrl(eventUrl)
+                    }
+
+                    if (isClick(event)) {
+                        this.clickCount += 1
+                    }
+
+                    if (isKeypress(event)) {
+                        this.keypressCount += 1
+                    }
+
+                    if (isMouseActivity(event)) {
+                        this.mouseActivityCount += 1
+                    }
+
+                    this.eventCount++
+                    this.size += chunk.length
+                }
+
+                rawBytesWritten += chunk.length
+            }
+        }
+
+        this.messageCount += 1
+        return rawBytesWritten
+    }
+```
+   - Accumulates events as newline-delimited JSON: `[windowId, event]\n[windowId, event]\n...`
+   - Tracks metadata: click_count, keypress_count, URLs, console logs, active_milliseconds
+   - Compresses each session block with Snappy
+4. **Flushes** periodically (max 10min buffer age or 50MB buffer size)
```

Comment:
it might have picked up blobby v1 flush triggers there (but i'm not certain by a long stretch)